### ISSUE TITLE
Clarification regarding NaN labels and model output dimensions

### ISSUE THREAD

**Issue by b-turan** [mlmed/torchxrayvision#119]
Hi there!

I am interested in training and evaluating a pretrained model on the cheXpert dataset. However, to me, it is not clear how to achieve this with the torchxrayvision library. In the following, I will provide a minimal working example to point out my confusion and am very happy about any suggestions:

```
import torch
import torchxrayvision as xrv

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Running on Device:", device)


# Load the pre-trained ResNet model
model = xrv.models.DenseNet(weights="densenet121-res224-chex", apply_sigmoid=False).to(device)

# CheXpert Dataset
dataset_chex = xrv.datasets.CheX_Dataset(
    imgpath=".data/CheXpert-v1.0-small/",
    csvpath=".data/CheXpert-v1.0-small/train.csv",
    transform=transforms.Compose(
        [
            xrv.datasets.XRayCenterCrop(),
            xrv.datasets.XRayResizer(224),
        ]
    ),
    unique_patients=False,
)

# Relabeling | Not absolutely clear how to deal with this functionality
xrv.datasets.relabel_dataset(xrv.datasets.default_pathologies, dataset_chex)
# Another option I consider: xrv.datasets.relabel_dataset(model.pathologies, dataset_chex)

test_loader = torch.utils.data.DataLoader(dataset_chex, batch_size=4, shuffle=False)

# Evaluate the model on the test data
with torch.no_grad():
    model.eval()
    # binary cross entropy loss
    criterion = torch.nn.BCEWithLogitsLoss()
    for sample in test_loader:
        images = sample["img"].to(device)  # (batch_size, 1, 224, 224)
        labels = sample["lab"].to(device)  # (batch_size, 18)
        # Get dimensions
        batch_size = images.shape[0]
        num_labels = torch.sum(~torch.isnan(labels[0]))
        # Inference model
        outputs = model(images)
        # Remove the labels that are not present in the dataset 
        outputs = outputs[~torch.isnan(labels)].view(batch_size, num_labels)
        labels = labels[~torch.isnan(labels)].view(batch_size, num_labels)
        # Compute the loss
        loss_sample = criterion(outputs, labels)
       # ... and so on
```

As I understand it, the cheXpert dataset considers 11 binary labels, while NaN values represent either non-existent labels or uncertain labels. The above pre-trained DenseNet has an output dimension of 18, with each output corresponding to an untrained label being 0.5, unlike NaN as stated in the publication.

Looking at the labels now, I am confused as to why NaN is used for both uncertain labels and untrained labels. Shouldn't they be separated so that one can distinguish between the two? Because of this confusion, I cannot simply reduce the labels to the appropriate classes that are important for the classification task or use another policy to deal with the uncertain labels.

I believe this is due to mishandling on my part, but if so, I would be grateful if you could tell me how to handle the data correctly. 

Thanks in Advance!

Best,
b-turan

**Comment by ieee8023** [comment]
> "densenet121-res224-chex"

This is not the official chexpert model. This is a model trained on the chexpert data. You can load the official chexpert model like this from the baseline models:
```python
model = xrv.baseline_models.chexpert.DenseNet(weights_zip="chexpert_weights.zip")
``` 

> with each output corresponding to an untrained label being 0.5, unlike NaN as stated in the publication.

The model ("densenet121-res224-chex") will output 0.5 for labels that it has not been trained on. The dataset will have label values of np.nan for unknown.

> NaN is used for both uncertain labels and untrained labels. Shouldn't they be separated so that one can distinguish between the two?

I think if the label is uncertain then that is not actionable information to train the model on anyway. The chexpert dataset just list -1 values for unknown. 

Are you referring to the pathologies not present in the chexpert data being nans and also that some labels for pathologies that are in the chexpert data are nans? The extra pathologies that are all nans are added by the relabel step. If you don't do that step then the those won't exist in the dataset. 

If you don't relabel the dataset then you will have to take a subset of the model's output that corresponds to what is in the dataset. An alternative is to not reduce the loss and then take a subset of the loss for the pathologies which matter. 

Why are you computing the loss anyway? Wouldn't you rather compute AUC values? This script may be what you are looking for? https://github.com/mlmed/torchxrayvision/blob/master/scripts/model_calibrate.py#L182 the outputs are saved for an entire dataset and then stored in a file and then read in and metrics are computed.

**Comment by b-turan** [comment]
Thanks for the quick reply!

> The model ("densenet121-res224-chex") will output 0.5 for labels that it has not been trained on. The dataset will have label values of np.nan for unknown.

Thanks for pointing that out, I was not sure about this.

> I think if the label is uncertain then that is not actionable information to train the model on anyway. The chexpert dataset just list -1 values for unknown.

As far as I know, there are several approaches for dealing with unknown labels as presented in the publication [cheXpert](https://arxiv.org/abs/1901.07031), such as ignore, set to 0, set to 1 and more. So I was confused that the uncertain labels are mixed with the non-existent labels. However, I will most likely switch to presenting uncertain labels as "0s", which is not a problem.

> Are you referring to the pathologies not present in the chexpert data being nans and also that some labels for pathologies that are in the chexpert data are nans? The extra pathologies that are all nans are added by the relabel step. If you don't do that step then the those won't exist in the dataset.

Exactly, but if I want to match the pathologies of the model with those of the dataset, I'd better call 
`xrv.datasets.relabel_dataset(model.pathologies, dataset_chex)` rather than `xrv.datasets.relabel_dataset(xrv.datasets.default_pathologies, dataset_chex)`, 
correct?

> Why are you computing the loss anyway? Wouldn't you rather compute AUC values? This script may be what you are looking for? https://github.com/mlmed/torchxrayvision/blob/master/scripts/model_calibrate.py#L182 the outputs are saved for an entire dataset and then stored in a file and then read in and metrics are computed.

You are right - the script was only for illustration of how I would handle the pre-processing and loading of the data. In the long run, I am interested in both _training_ and _evaluating_ a novel XAI method on the cheXpert dataset. For this purpose, I need a pre-trained classifier that is then re-trained with two additional agents. As I continue with the work, I may ask additional questions about the library and would be happy for your support.

**Comment by ieee8023** [comment]
> xrv.datasets.relabel_dataset(model.pathologies, dataset_chex)

Yes, this will align the dataset to the model. It will insert extra pathologies in the dataset with a value of np.nan.

> I am interested in both training and evaluating

The way the loss is computed during training for the pre-trained models is here:

https://github.com/mlmed/torchxrayvision/blob/11c53a046ea7335cf3b132d61c22af81290748d5/scripts/train_utils.py#L174-L185

Also here is the main training script that will train and report validation metrics during training in order to early stop: https://github.com/mlmed/torchxrayvision/blob/11c53a046ea7335cf3b132d61c22af81290748d5/scripts/train_utils.py

If I was to do it over I would use PyTorch Lightning for the training and early stopping.

> I may ask additional questions about the library and would be happy for your support.

Of course!

**Comment by ieee8023** [comment]
Let me know if this issue is not resolved. I'll close it for now.
